File: Backend/src/scud/crypto/sealed_box.py
# X25519 + ChaCha20-Poly1305-IETF sealed box decryption.
# Follows shared §2.4: nonce_24 = BLAKE2b(ephemeral_pub || server_x25519_pub, 24),
# then chacha nonce = nonce_24[:12] (IETF variant uses 12-byte nonce).

import hashlib

from nacl.bindings import crypto_scalarmult
from nacl.bindings.crypto_aead import crypto_aead_chacha20poly1305_ietf_decrypt


def decrypt_sealed_blob(
    blob: bytes,
    server_x25519_priv: bytes,
    server_x25519_pub: bytes,
) -> bytes:
    """Decrypt a sealed blob produced by the reader.

    blob layout: ephemeral_pub(32) || ciphertext+tag(N+16)
    """
    if len(blob) < 48:
        raise ValueError(f"sealed blob too short: {len(blob)} B")

    ephemeral_pub = blob[:32]
    ct_and_tag = blob[32:]

    # X25519 key exchange
    shared = crypto_scalarmult(server_x25519_priv, ephemeral_pub)

    # KDF (CRYPTO-05 "blender"): derive a uniform AEAD key from the raw X25519
    # secret via BLAKE2b(shared, 32). MUST match firmware-encrypt exactly.
    key = hashlib.blake2b(shared, digest_size=32).digest()

    # Derive 12-byte nonce from BLAKE2b(ephemeral_pub || server_pub, 24)
    nonce_24 = hashlib.blake2b(
        ephemeral_pub + server_x25519_pub, digest_size=24
    ).digest()
    nonce_12 = nonce_24[:12]

    return crypto_aead_chacha20poly1305_ietf_decrypt(
        ct_and_tag,
        aad=None,
        nonce=nonce_12,
        key=key,
    )


def encrypt_sealed_blob(
    plaintext: bytes,
    server_x25519_pub: bytes,
) -> bytes:
    """Encrypt plaintext for server — used in tests as reference."""
    from nacl.bindings.crypto_aead import crypto_aead_chacha20poly1305_ietf_encrypt
    from nacl.public import PrivateKey

    ephemeral_key = PrivateKey.generate()
    ephemeral_priv = bytes(ephemeral_key)
    ephemeral_pub = bytes(ephemeral_key.public_key)

    shared = crypto_scalarmult(ephemeral_priv, server_x25519_pub)

    # KDF (CRYPTO-05 "blender"): same as decrypt — BLAKE2b(shared, 32).
    key = hashlib.blake2b(shared, digest_size=32).digest()

    nonce_24 = hashlib.blake2b(
        ephemeral_pub + server_x25519_pub, digest_size=24
    ).digest()
    nonce_12 = nonce_24[:12]

    ct_and_tag = crypto_aead_chacha20poly1305_ietf_encrypt(
        plaintext,
        aad=None,
        nonce=nonce_12,
        key=key,
    )
    return ephemeral_pub + ct_and_tag

File: ESP32/firmware/src/crypto/sealed_box.cpp
#include "sealed_box.h"
#include <string.h>
#include "monocypher.h"

// X25519 + IETF ChaCha20-Poly1305 (RFC 8439, 12-B nonce) — всё на Monocypher.
// mbedTLS в Arduino-сборке ESP32 собран без MBEDTLS_CHACHAPOLY_C, поэтому
// AEAD реализован вручную из crypto_chacha20_ietf + crypto_poly1305.
// Monocypher'овский crypto_aead_lock нельзя — у него XChaCha20 (24-B nonce).

static void u64_le(uint8_t out[8], uint64_t v) {
    for (int i = 0; i < 8; ++i) out[i] = (uint8_t)(v >> (8 * i));
}

bool sealed_box_encrypt(const uint8_t server_x25519_pub[32],
                        const uint8_t* plaintext, size_t pt_len,
                        uint8_t* out_blob,
                        const uint8_t random_32[32]) {
    uint8_t ephemeral_priv[32];
    memcpy(ephemeral_priv, random_32, 32);

    uint8_t* ephemeral_pub = out_blob;
    crypto_x25519_public_key(ephemeral_pub, ephemeral_priv);

    uint8_t shared[32];
    crypto_x25519(shared, ephemeral_priv, server_x25519_pub);
    crypto_wipe(ephemeral_priv, 32);

    // KDF (CRYPTO-05 "blender"): derive a uniform AEAD key from the raw X25519
    // secret via BLAKE2b(shared, 32). MUST match backend-decrypt exactly.
    uint8_t key[32];
    crypto_blake2b(key, 32, shared, 32);
    crypto_wipe(shared, 32);

    uint8_t nonce24[24];
    crypto_blake2b_ctx ctx;
    crypto_blake2b_init(&ctx, 24);
    crypto_blake2b_update(&ctx, ephemeral_pub, 32);
    crypto_blake2b_update(&ctx, server_x25519_pub, 32);
    crypto_blake2b_final(&ctx, nonce24);

    uint8_t nonce12[12];
    memcpy(nonce12, nonce24, 12);

    uint8_t* ct  = out_blob + 32;
    uint8_t* tag = out_blob + 32 + pt_len;

    // Poly1305 one-time key = ChaCha20(key, nonce12, counter=0)[:32].
    uint8_t zero64[64] = {0};
    uint8_t keystream0[64];
    crypto_chacha20_ietf(keystream0, zero64, 64, key, nonce12, 0);
    uint8_t poly_key[32];
    memcpy(poly_key, keystream0, 32);
    crypto_wipe(keystream0, 64);

    // Encrypt starting at counter=1.
    crypto_chacha20_ietf(ct, plaintext, pt_len, key, nonce12, 1);
    crypto_wipe(key, 32);

    // MAC input (AAD пустой): ct || pad16(ct) || len(aad)_le64 || len(ct)_le64.
    crypto_poly1305_ctx pctx;
    crypto_poly1305_init(&pctx, poly_key);
    crypto_poly1305_update(&pctx, ct, pt_len);
    size_t pad_ct = (16 - (pt_len % 16)) % 16;
    if (pad_ct) {
        uint8_t zeros[16] = {0};
        crypto_poly1305_update(&pctx, zeros, pad_ct);
    }
    uint8_t lens[16] = {0};
    u64_le(lens + 0, 0);
    u64_le(lens + 8, (uint64_t)pt_len);
    crypto_poly1305_update(&pctx, lens, 16);
    crypto_poly1305_final(&pctx, tag);
    crypto_wipe(poly_key, 32);

    return true;
}
